File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @day_seconds (24 * 60 * 60)
  @year_days %{
    earth: 365.26,
    jupiter: 4_332.82,
    mars: 686.98,
    mercury: 87.97,
    neptune: 60_190.03,
    saturn: 10_755.70,
    uranus: 30_687.15,
    venus: 224.7
  }

  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
  | :saturn | :neptune | :uranus

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds) do
    seconds / Map.get(@year_days, planet) / @day_seconds
  end
end

File: javascript/space-age/space-age.js
'use strict';

const PLANET_YEAR_DAYS = {
  Earth: 365.26,
  Jupiter: 4332.82,
  Mars: 686.98,
  Mercury: 87.97,
  Neptune: 60190.03,
  Saturn: 10755.70,
  Uranus: 30687.15,
  Venus: 224.7
};
const PLANETS = Object.keys(PLANET_YEAR_DAYS);
const DAY_SECONDS = 24 * 60 * 60;

class SpaceAge {
  constructor(seconds) {
    this.seconds = seconds;
  }
}

PLANETS.forEach((planet) => {
  let functionName = 'on' + planet;
  SpaceAge.prototype[functionName] = function() {
    return onPlanet(this.seconds, planet);
  }
});

function onPlanet(seconds, planet) {
  return round(seconds / (DAY_SECONDS * PLANET_YEAR_DAYS[planet]));
}

function round(float) {
  return Math.round(float * 100) / 100;
}

module.exports = SpaceAge;
